hefquin-engine/src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpBindJoinSPARQLwithVALUESorFILTER.java

```diff
@@ -0,0 +1,84 @@
+import se.liu.ida.hefquin.engine.queryplan.executable.IntermediateResultElementSink;
+import se.liu.ida.hefquin.engine.queryproc.ExecutionContext;
+
+/**
+ * Implementation of (a batching version of) the bind join algorithm that starts by
+ * using a VALUES clause exactly as done by {@link ExecOpBindJoinSPARQLwithVALUES}.
+ * If this fails for the first request, the implementation repeats the request by
+ * using FILTERs as done by {@link ExecOpBindJoinSPARQLwithFILTER} and, then,
+ * continues using the FILTER-based approach for the rest of the requests. If the
+ * first VALUES-based request succeeds, however, then the implementation continues
+ * using the VALUES-based approach for the rest of the requests. 
+ */
+public class ExecOpBindJoinSPARQLwithVALUESorFILTER extends BaseForExecOpBindJoin<SPARQLGraphPattern, SPARQLEndpoint> {
+
+	// will be initialized when processing the first input block of solution mappings
+	protected BaseForExecOpBindJoinSPARQL currentInstance = null;
+	boolean valuesBasedRequestFailed;
+	
+	
+	public ExecOpBindJoinSPARQLwithVALUESorFILTER( final TriplePattern query, 
+	                                               final SPARQLEndpoint fm,
+	                                               final boolean useOuterJoinSemantics,
+	                                               final boolean collectExceptions ) {
+		super(query, fm, collectExceptions);
+	}
+
+
+	@Override
+	protected void _process( final IntermediateResultBlock input, final IntermediateResultElementSink sink, final ExecutionContext execCxt )
+			throws ExecOpExecutionException {
+		
+		//If this is the first request
+		if (currentInstance == null) {
+			currentInstance = new ExecOpBindJoinSPARQLwithVALUES(query, fm, collectExceptions);
+			valuesBasedRequestFailed = false;
+			try {
+				// Try using VALUES-based bind join
+				currentInstance.process(input, sink, execCxt);
+				if (!currentInstance.getExceptionsCaughtDuringExecution().isEmpty()) {
+					valuesBasedRequestFailed = true;
+				}	
+			} catch ( final ExecOpExecutionException e ) {
+				valuesBasedRequestFailed = true;
+			}
+			if (valuesBasedRequestFailed == true) {
+				// Use FILTER-based bind join instead
+				currentInstance = new ExecOpBindJoinSPARQLwithFILTER(query, fm, currentInstance.useOuterJoinSemantics, collectExceptions);
+				currentInstance.process(input, sink, execCxt);
+			}
+		}
+		else {
+			currentInstance.process(input, sink, execCxt);
+		}
+	}
+
+	
+	@Override
+	public List<Exception> getExceptionsCaughtDuringExecution() {
+		return currentInstance.getExceptionsCaughtDuringExecution();
+	}
+
+
+    @Override
+    protected void _concludeExecution(
+            final IntermediateResultElementSink sink,
+            final ExecutionContext execCxt ) {
+    	if (currentInstance!= null) {
+    		 try {
+				currentInstance.concludeExecution(sink, execCxt);
+			} catch (ExecOpExecutionException e) {
+				e.printStackTrace();
+			}
+    	}
+    }
```

Comment:
Printing the stack trace of a potential exception and not actually doing anything about the exception is not really robust code ;-)   In this case, the potential exceptions don't need to be caught but can simply be passed on by this function. Here is how this would look like (plus some code layout changes):
```suggestion
	@Override
	protected void _concludeExecution( final IntermediateResultElementSink sink,
	                                   final ExecutionContext execCxt ) throws ExecOpExecutionException {
		if ( currentInstance != null ) {
			currentInstance.concludeExecution(sink, execCxt);
		}
	}
```